packages/theme/src/cli/utilities/theme-environment/proxy.ts

```diff
@@ -0,0 +1,185 @@
+  const mainCdnRE = /(?:https?:)?\/\/cdn\.shopify\.com\/(.*?\/(assets\/[^?">]+)(?:\?|"|>|$))/g
+  const existingAssets = new Set([...ctx.localThemeFileSystem.files.keys()].filter((key) => key.startsWith('assets')))
+  content = content.replace(mainCdnRE, (matchedUrl, pathname, matchedAsset) => {
+    const isLocalAsset = matchedAsset && existingAssets.has(matchedAsset as string)
+    // Prefix with vanityCdnPath to later read local assets
+    return isLocalAsset ? `${vanityCdnPath}${pathname}` : matchedUrl
+  })
+
+  return content
+}
+
+function patchBaseUrlAttributes(html: string, ctx: DevServerContext) {
+  const newBaseUrl = `http://${ctx.options.host}:${ctx.options.port}`
+  const dataBaseUrlRE = new RegExp(`data-base-url=["']((?:https?:)?//${getStoreFqdnForRegEx(ctx)})[^"']*?["']`, 'g')
+
+  return html.replace(dataBaseUrlRE, (match, m1) => match.replace(m1, newBaseUrl))
+}
+
+function patchCookieDomains(cookieHeader: string[], ctx: DevServerContext) {
+  // Domains are invalid for localhost:
+  const domainRE = new RegExp(`Domain=${getStoreFqdnForRegEx(ctx)};\\s*`, 'gi')
+  return cookieHeader.map((value) => value.replace(domainRE, '')).join(', ')
+}
+
+/**
+ * Patches the result of an SFR HTML response to include the local proxies
+ * and fix domain inconsistencies between remote instance and local dev.
+ */
+export async function patchRenderingResponse(ctx: DevServerContext, event: H3Event, response: NodeResponse) {
+  setResponseHeaders(event, Object.fromEntries(response.headers.entries()))
+  patchProxiedResponseHeaders(ctx, event, response)
+
+  // We are decoding the payload here, remove the header:
+  let html = await response.text()
+  removeResponseHeader(event, 'content-encoding')
+
+  html = injectCdnProxy(html, ctx)
+  html = patchBaseUrlAttributes(html, ctx)
+
+  return html
+}
+
+// These headers are meaningful only for a single transport-level connection,
+// and must not be retransmitted by proxies or cached.
+// https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-14#section-7.1.3.1Acc
+const HOP_BY_HOP_HEADERS = [
+  'connection',
+  'keep-alive',
+  'proxy-authenticate',
+  'proxy-authorization',
+  'te',
+  'trailer',
+  'transfer-encoding',
+  'upgrade',
+  'content-security-policy',
+  'content-length',
+]
+
+function patchProxiedResponseHeaders(ctx: DevServerContext, event: H3Event, response: Response | NodeResponse) {
+  // Safari adds upgrade-insecure-requests to CSP and it needs to be removed:
+  clearResponseHeaders(event, HOP_BY_HOP_HEADERS)
+
+  // Link header preloads resources from global CDN, proxy it:
+  const linkHeader = response.headers.get('Link')
+  if (linkHeader) setResponseHeader(event, 'Link', injectCdnProxy(linkHeader, ctx))
+
+  // Cookies are set for the vanity domain, fix it for localhost:
+  const setCookieHeader =
+    'raw' in response.headers ? response.headers.raw()['set-cookie'] : response.headers.getSetCookie()
```

Comment:
Note: H3 uses native `fetch` and `Response`, while cli-kit's fetch utilities are based on `node-fetch`. This means that the result of H3's `sendProxy` is a Response, whereas the result of our `render` function is a NodeResponse. The main difference is in the way we access `set-cookie` headers. They also differs in `response.body` stream implementation but we don't use that here (`response.text()` work in both).